ports-of-call/span.hpp

```diff
@@ -0,0 +1,464 @@
+template <class E>
+struct span_storage<E, dynamic_extent> {
+  constexpr span_storage() noexcept = default;
+
+  constexpr span_storage(E *p_ptr, std::size_t p_size) noexcept
+      : ptr(p_ptr), size(p_size) {}
+
+  E *ptr = nullptr;
+  std::size_t size = 0;
+};
+
+// some metaprogramming
+template <class T>
+using uncvref_t = typename std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+// detector of span
+template <class Q>
+struct is_span_oracle : std::false_type {};
+
+template <class T, std::size_t Extent>
+struct is_span_oracle<span<T, Extent>> : std::true_type {};
+
+template <class Q>
+struct is_span : is_span_oracle<typename std::remove_cv<Q>::type> {};
+
+template <class Q>
+struct is_std_array_oracle : std::false_type {};
+
+template <class T, std::size_t Extent>
+struct is_std_array_oracle<std::array<T, Extent>> : std::true_type {};
+
+template <class Q>
+struct is_std_array : is_std_array_oracle<typename std::remove_cv<Q>::type> {};
+
+template <class Q>
+struct is_array : std::false_type {};
+
+template <class T>
+struct is_array<T[]> : std::true_type {};
+
+template <class T, std::size_t N>
+struct is_array<T[N]> : std::true_type {};
+
+// detect C.size() and C.data()
+template <class, class = void>
+struct has_size_and_data : std::false_type {};
+
+template <class C>
+struct has_size_and_data<C, std::void_t<decltype(std::size(std::declval<C>())),
+                                        decltype(std::data(std::declval<C>()))>>
+    : std::true_type {};
+
+// detect if *(C.data()) is compatible with E
+template <class, class, class = void>
+struct is_compatible_element : std::false_type {};
+
+template <class C, class E>
+struct is_compatible_element<C, E, std::void_t<decltype(std::data(std::declval<C>()))>>
+    : std::is_convertible<typename std::remove_pointer<decltype(std::data(
+                              std::declval<C &>()))>::type (*)[],
+                          E (*)[]> {};
+
+// A container C:
+// - is not any of [span, c-array, std::array]
+// - has C.size() and C.data() interface
+template <class C>
+struct is_container
+    : std::bool_constant<!is_span<C>::value && !is_array<C>::value &&
+                         !is_std_array<C>::value && has_size_and_data<C>::value> {};
+
+// detect if C is a container and is compatible with element E
+template <class C, class E>
+struct is_compatible_container
+    : std::bool_constant<is_container<C>::value && is_compatible_element<C, E>::value> {};
+
+// detect if T is a complete type
+// NB: sizeof(T) cannot be used with incomplete types
+template <class, class = std::size_t>
+struct is_complete : std::false_type {};
+
+template <class T>
+struct is_complete<T, decltype(sizeof(T))> : std::true_type {};
+
+} // namespace detail
+
+template <class ElementType, std::size_t Extent>
+class span {
+  static_assert(std::is_object<ElementType>::value,
+                "A span's ElementType must be an object type (not a "
+                "reference type or void)");
```

Comment:
And why is this?

Reply:
contractually, `span` provides `size_bytes() { return size() * sizeof(element_type)`. `sizeof(void)` is undefined, and `sizeof(T&)` is in general incorrect. Further, `span` defines `reference_type = element_type &`, and things can get confusing very quickly. 

Disallowing references is a deeper topic, but basically has to do with how associated iterator types are determined, see https://en.cppreference.com/w/cpp/iterator/iter_t for a fun time